lib/vagrant-service-manager/adb_executor.rb

```diff
@@ -0,0 +1,40 @@
+module VagrantPlugins
+  module ServiceManager
+    VERSION_CMD = {
+      docker: "docker version --format '{{.Server.Version}}'",
+      openshift: "oc version | grep 'oc' | grep -oE '[0-9.]+'"
+    }
+
+    class ADBExecutor
+      LABEL = 'servicemanager.commands.install_cli.message'
+
+      def initialize(binary_type, machine, ui, options)
+        @type = binary_type
+        @machine = machine
+        @ui = ui
+        @options = options
+      end
+
+      def execute
+        binary_exists = true
+        version = @options['--cli-version']
+        path = @options['--path']
+
+        version = PluginUtil.execute_once(@machine, @ui, VERSION_CMD[@type.to_sym]) if version.nil?
+        path = PluginUtil.service_class(@type).binary_path(version) if path.nil?
+
+        unless PluginUtil.binary_downloaded?(path)
+          binary_exists = false
+          url = PluginUtil.service_class(@type).binary_download_url(@machine, version)
+          PluginUtil.download_binary(@type, url, path)
+        end
+
+        binary_name = @type
```

Comment:
> Actually, binary name is different then the type like in case of openshift it is oc and kubectl in case of kubernetes but for docker it is docker.

I understand what you are doing, but it is not really easy to understand. I think there are several things which could improve that. The easiest being a dedicated method with a descriptive name/. Or maybe one of the helper methods above can return the name?

> May be we can pull this code into PluginUtil.

I am not a big fan of this idea. As said, `PluginUtil` is imo already too big and dealing with too many different aspects.